server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java

```diff
@@ -127,17 +132,15 @@ void runIfNotProcessed(BatchedTask updateTask) {
         if (updateTask.processed.get() == false) {
             final List<BatchedTask> toExecute = new ArrayList<>();
             final Map<String, List<BatchedTask>> processTasksBySource = new HashMap<>();
-            synchronized (tasksPerBatchingKey) {
-                LinkedHashSet<BatchedTask> pending = tasksPerBatchingKey.remove(updateTask.batchingKey);
-                if (pending != null) {
-                    for (BatchedTask task : pending) {
-                        if (task.processed.getAndSet(true) == false) {
-                            logger.trace("will process {}", task);
-                            toExecute.add(task);
-                            processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
-                        } else {
-                            logger.trace("skipping {}, already processed", task);
-                        }
+            final Set<BatchedTask> pending = tasksPerBatchingKey.remove(updateTask.batchingKey);
+            if (pending != null) {
+                for (BatchedTask task : pending) {
```

Comment:
I think we would want to synchronize explicitly on the `pending` set instead of relying on `Collections.synchronizedSet`, iterating a synchronized set like this is not really safe.

Reply:
Right fair point :) Added locking to this loop.